Feature: Added Wallaby support for skeleton-typescript-webpack

pndewit opened the pull request:
New feature for skeleton-typescript-webpack: Added Wallaby support.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aurelia/skeleton-navigation?pullRequest=722) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aurelia/skeleton-navigation?pullRequest=722) before we can accept your contribution.<br/><hr/>**Patrick de Wit** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

niieani commented:
We do not have a Wallaby license. We will welcome a contribution, but have no way of doing this ourselves.

niieani commented on skeleton-typescript-webpack/package.json:
```diff
@@ -113,6 +113,7 @@
     "@types/webpack": "^2.0.0",
     "@types/whatwg-fetch": "^0.0.32",
     "aurelia-protractor-plugin": "^1.0.1",
+    "babel-polyfill": "^6.16.0",
```
Why is the babel-polyfill needed for the typescript skeleton to work?

Also, can you make the same change to the esnext skeleton?

pndewit replied:
The "wallaby-webpack" module uses Promise. Without the babel-polyfill dependency it gives me this error:
```
1 failing tests, 9 passing

test/unit/app.spec.ts the App module contains a router property [5 ms]

	ReferenceError: Can't find variable: Promise
		at src/users.ts:5

	ReferenceError: Can't find variable: Promise
	at http://localhost:65114/wallaby-webpack.js?1480348701747:1
```

I can try to also add it to the esnext skeleton.

niieani replied:
That's so odd. Which Node.js version are you on? Shouldn't it have a Promise implementation? In any case, we could use Bluebird instead of Babel, since it's already there as a dependency (no need for us to have multiple Promise implementations).

pndewit replied:
I am using Node.js version 6.2, but it's probably because Wallaby runs a Phantom browser and afaik Phantom does not support the Promise API (yet), right?
You are so right about the Bluebird implementation of Promise. Updated my commit!

pndewit commented:
As requested, also added support for Wallaby in skeleton-esnext-webpack. For some reason I couldn't get it to work there with bluebird inside the wallaby config..? I was constantly getting errors that Map could not be found. For that reason I added the babel-polyfill dependency.
If someone knows a fix for this, I would be happy to implement it and remove the babel-polyfill again!

niieani commented:
Thanks. I'd hold off with merging this until we can get it to work using a single Promises implementation, since we don't want to bloat the skeleton.

pndewit commented:
@niieani I am not bloating the skeleton for the skeleton-typescript-webpack with extra Promises implementations. I removed the unnecessary one.

For the skeleton-esnext-webpack skeleton I am adding an extra implementation, but not for Promise, but for Map. If you don't want to merge this PR because of this, I will split it into two PR's. One for skeleton-typescript-webpack and one for skeleton-esnext-webpack.

niieani commented:
@pndewit A polyfill for `Map` is present in `aurelia-polyfills`, why not use that one instead?

pndewit commented:
@niieani I tried, but that didn't work either. `aurelia-polyfills` doesn't work out of the box because Wallaby doesn't understand ES modules (yet). If I try to run it with `aurelia-polyfills` instead of `babel-polyfill`, it starts complaining about `import`.

niieani commented:
@pndewit seems to me you're not referencing the compiled files. There are no `import` references or ES modules if you pick, say, the commonjs module: https://github.com/aurelia/polyfills/tree/master/dist

pndewit commented:
@niieani None of those are browser compatible.
https://github.com/aurelia/polyfills/blob/master/dist/aurelia-polyfills.js uses import.
https://github.com/aurelia/polyfills/blob/master/dist/commonjs/aurelia-polyfills.js uses require.
https://github.com/aurelia/polyfills/blob/master/dist/amd/aurelia-polyfills.js uses define.
https://github.com/aurelia/polyfills/blob/master/dist/es2015/aurelia-polyfills.js uses import.
https://github.com/aurelia/polyfills/blob/master/dist/native-modules/aurelia-polyfills.js uses import.
https://github.com/aurelia/polyfills/blob/master/dist/system/aurelia-polyfills.js uses System.

niieani commented:
Hold on, why would you even want to load this manually? The `aurelia-pollyfills` module should be loaded through the Webpack entrypoint. Seems like something's wrong with the configuration of Wallaby (in that it's not using the Webpack config). I'm no Wallaby specialist, but aren't you supposed to pass the config to the wallabyWebpack call as a parameter?

pndewit commented:
@niieani That is true! I have also tried this using: https://wallabyjs.com/docs/integration/webpack.html
But the webpack config in the skeleton is written with `import` statements as well. It will first need to be transpiled. This can be done with babel runtime I guess. But I don't think this one is added as dependency either.

niieani commented:
I'm ok with adding babel-node (compile in-place), the typescript skel already has `ts-node` which is the equivalent. 

It should be available in `babel-cli`. See: https://babeljs.io/docs/usage/cli/#babel-node

pndewit commented:
@niieani I tried to do this with `babel-register` (https://babeljs.io/docs/usage/babel-register/) since I found this line interesting:
`All subsequent files required by node with the extensions .es6, .es, .jsx and .js will be transformed by Babel. The polyfill specified in polyfill is also automatically required.`.

Unfortunately it gives me a parsing error. I used the following config to transpile the webpack config:
```
require('babel-register')({
    presets: ['es2015', 'stage-1'],
    only: './webpack.config.babel'
});
var webpackConfig = require('./webpack.config.babel');
```

It gives me the following error: `SyntaxError: Unexpected token ! in JSON at position 0`

niieani commented:
`only` parameter has to be a Regex. You provided a string. Try `only: /webpack\.config\.babel\.js$`

niieani commented:
Or just do:

`extensions: [".babel.js"]`

pndewit commented:
Both can't find the file as it results into the error that it doesn't understand `import`

niieani commented:
so this:
`extensions: [".babel.js"]`
and this:
`require('./webpack.config');`
doesn't work?

pndewit commented:
My bad, extensions: [".babel.js"] can't find the file. only: /webpack\.config\.babel\.js$ does, but gives me the same: SyntaxError: Unexpected token ! in JSON at position 0 error

niieani commented:
I'm sorry, but I can't help you debug this anymore, it's taking too much time. If someone else would like to chime in, please go ahead.

pndewit commented:
Shall I just split the two skel commits into two PR's? So that the `skeleton-typescript-webpack` PR can be merged?

niieani commented:
Shouldn't both skeletons load the proper wallaby config? You'd need to use `ts-node/register` to do that.

pndewit commented:
I don't think I completely understand what you are saying.

"Should" both skeletons load a wallaby config? Yes. (support for wallaby in each skel would be nice)
Do both skeletons "need" to load a wallaby config? No. (they do not depend on each other to work right?)

pndewit commented:
Did some digging and found out it had something to do with babel-register. See:
https://github.com/wallabyjs/public/issues/726
https://github.com/wallabyjs/public/issues/863

Adding the following piece fixed the error: `SyntaxError: Unexpected token ! in JSON at position 0`:

```
var Module              = require( 'module' ).Module;
var modulePrototype     = Module.prototype;
var originalRequire     = modulePrototype.require;
modulePrototype.require = function ( filePath ) {
  if ( filePath === 'source-map-support' ) {
    return { install: () => {} };
  }
  return originalRequire.call( this, filePath );
};

require('babel-register')({
  presets: ['es2015', 'stage-1'],
  only: /webpack\.config\.babel\.js$/
});

modulePrototype.require = originalRequire;
process.env.NODE_ENV = 'test';

var wallabyWebpack = require('wallaby-webpack');
var webpackConfig = require('./webpack.config.babel');
var wallabyPostprocessor = wallabyWebpack(webpackConfig);
```

Now Wallaby crashes on this:

```
Failed to parse src/blur-image.js, SyntaxError: Unexpected character '@' (3:0)
  at emitTwo (events.js:106:13)
  at process.emit (events.js:191:7)
  at handleMessage (internal/child_process.js:718:10)
  at Pipe.channel.onread (internal/child_process.js:444:11)
  at _combinedTickCallback (internal/process/next_tick.js:67:7)
  at process._tickCallback (internal/process/next_tick.js:98:9)
```

Is Webpack not doing its work or something...?

pndewit commented:
Doing a transpile in the wallaby config as well:

```
...
    compilers: {
      'src/**/*.js': wallaby.compilers.babel({
        presets: ['es2015', 'stage-1'],
        plugins: [
          'syntax-flow',
          'transform-decorators-legacy',
          'transform-flow-strip-types'
        ]
      })
    },
...
```

Is giving the following error:
```
No context elements
No context elements
TypeError: Cannot convert undefined or null to object
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/aurelia-webpack-plugin/dist/index.js:194:58
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/lib/ContextModuleFactory.js:151:5
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/node_modules/async/dist/async.js:1013:9
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/node_modules/async/dist/async.js:356:16
  at iteratorCallback (/Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/node_modules/async/dist/async.js:936:13)
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/node_modules/async/dist/async.js:840:16
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/node_modules/async/dist/async.js:1010:13
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/lib/ContextModuleFactory.js:139:8
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/tapable/lib/Tapable.js:131:11
  at ContextModuleFactory.<anonymous> (/Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/lib/dependencies/RequireContextPlugin.js:62:4)
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/tapable/lib/Tapable.js:133:14
  at ContextModuleFactory.<anonymous> (/Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/lib/dependencies/RequireContextPlugin.js:43:4)
  at ContextModuleFactory.applyPluginsAsyncWaterfall (/Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/tapable/lib/Tapable.js:135:13)
  at ContextModuleFactory.<anonymous> (/Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/webpack/lib/ContextModuleFactory.js:130:12)
  at /Users/patrickdewit/projects/skeleton-navigation/skeleton-esnext-webpack/node_modules/graceful-fs/polyfills.js:287:18
  at FSReqWrap.oncomplete (fs.js:117:15)

No context elements
No context elements
No context elements
```

niieani commented:
@pndewit thanks for digging in. Seems like somethings not going right with aurelia-webpack-plugin here. But it's not a big deal, since we'll be deprecating that soon. It might be easier to add the Wallaby support changes to the new experimental skeleton once that is released.

pndewit commented:
@niieani That would explain it. Alright, sounds good to me! I will revert the changes I made to the skeleton-esnext-webpack skeleton. Hope the changes for the skeleton-typescript-webpack can be merged after! :)

pndewit commented:
It should be ready for merging. Let me know if I am missing something.
